quickwit-directories/src/union_directory.rs

```diff
@@ -0,0 +1,161 @@
+
+/// A union directory takes a bunch of directory and stack them similarly to UnionFS.
+/// The resulting directory is a virtual view of the union of the different directories.
+///
+/// If a path exists in all directories, the first of the list containing the path
+/// will shadow the other directories.
+///
+/// The first directory of the list will receive all write operations.
+/// Delete on the other hand will be applied on all directories containing the file.
+#[derive(Debug, Clone)]
+pub struct UnionDirectory {
+    directories: Arc<Vec<Box<dyn Directory>>>,
+}
+
+impl UnionDirectory {
+    /// Creates a new union directory.
+    pub fn union_of(directories: Vec<Box<dyn Directory>>) -> UnionDirectory {
+        UnionDirectory {
+            directories: Arc::new(directories),
+        }
+    }
+
+    /// Helper function to find the first directory containing the given path.
+    fn find_directory_for_path(&self, path: &Path) -> Result<&dyn Directory, OpenReadError> {
+        for directory in self.directories.iter() {
+            if directory.exists(path)? {
+                return Ok(directory.as_ref());
+            }
+        }
+        Err(OpenReadError::FileDoesNotExist(path.to_path_buf()))
+    }
+}
+
+impl Directory for UnionDirectory {
+    fn get_file_handle(&self, path: &Path) -> Result<Box<dyn FileHandle>, OpenReadError> {
+        let directory = self.find_directory_for_path(path)?;
+        directory.get_file_handle(path)
+    }
+
+    fn exists(&self, path: &Path) -> Result<bool, OpenReadError> {
+        match self.find_directory_for_path(path) {
+            Ok(_) => Ok(true),
+            Err(OpenReadError::FileDoesNotExist(_)) => Ok(false),
+            Err(err) => Err(err),
+        }
+    }
+
+    fn atomic_read(&self, path: &Path) -> Result<Vec<u8>, OpenReadError> {
+        let directory = self.find_directory_for_path(path)?;
+        directory.atomic_read(path)
+    }
+
+    fn open_write(&self, path: &Path) -> Result<tantivy::directory::WritePtr, OpenWriteError> {
+        self.directories[0].open_write(path)
+    }
+
+    fn delete(&self, path: &Path) -> Result<(), DeleteError> {
+        let mut deleted_at_least_one_file = false;
+        for directory in self.directories.iter() {
+            match directory.delete(path) {
+                Ok(()) => {
+                    deleted_at_least_one_file = true;
+                }
+                Err(DeleteError::FileDoesNotExist(_)) => {}
+                err => {
+                    return err;
+                }
+            }
+        }
+        if deleted_at_least_one_file {
+            Ok(())
+        } else {
+            Err(DeleteError::FileDoesNotExist(path.to_path_buf()))
+        }
+    }
+
+    fn atomic_write(&self, path: &Path, data: &[u8]) -> io::Result<()> {
+        self.directories[0].atomic_write(path, data)
+    }
+
+    fn watch(&self, _: tantivy::directory::WatchCallback) -> tantivy::Result<WatchHandle> {
+        Ok(WatchHandle::empty())
```

Comment:
Should we prefer `unimplemented!("``UnionDirectory`` does not support watch.")`?

Reply:
I think this is ok (and required).